File: 1558-course-schedule-iv/course-schedule-iv.java
import java.util.*;

class Solution {
    public List<Boolean> checkIfPrerequisite(int n, int[][] prerequisites, int[][] queries) {
        // Build adjacency list
        List<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        for (int[] edge : prerequisites) {
            int u = edge[0], v = edge[1];
            adj[u].add(v);
        }

        // Initialize reachability matrix
        boolean[][] reachable = new boolean[n][n];

        // Perform BFS for each node to find all reachable nodes
        for (int i = 0; i < n; i++) {
            Queue<Integer> queue = new LinkedList<>();
            queue.offer(i);
            while (!queue.isEmpty()) {
                int u = queue.poll();
                for (int v : adj[u]) {
                    if (!reachable[i][v]) {
                        reachable[i][v] = true;
                        queue.offer(v);
                    }
                }
            }
        }

        // Answer the queries
        List<Boolean> ans = new ArrayList<>();
        for (int[] query : queries) {
            int u = query[0], v = query[1];
            ans.add(reachable[u][v]);
        }

        return ans;
    }
}

File: 1558-course-schedule-iv/course-schedule-iv.py
class Solution:
    def checkIfPrerequisite(self, n: int, p: List[List[int]], q: List[List[int]]) -> List[bool]:
        reachable = [[False for _ in range(n)] for _ in range(n)]
        adj = [[] for _ in range(n)]
        for u, v in p:
            adj[u].append(v)
        
         # Perform BFS for each node to find all reachable nodes
        for i in range(n):
            queue = deque([i])
            while queue:
                u = queue.popleft()
                for v in adj[u]:
                    if not reachable[i][v]:
                        reachable[i][v] = True
                        queue.append(v)
        
        # Answer the queries
        ans = []
        for u, v in q:
            ans.append(reachable[u][v])
        return ans


